File: app/src/main/java/com/example/engine/HaplogroupClassifier.kt
package com.example.engine

import com.example.data.db.HaplogroupTreeData
import com.example.data.db.SnpReferenceDatabase
import com.example.data.model.DnaAnalysisResult
import com.example.data.model.EvaluatedMarker
import com.example.data.model.HaplogroupDefinition
import com.example.data.model.LineageAnalysis
import com.example.data.model.LineageType
import com.example.data.model.MarkerStatus
import com.example.data.model.SnpMarker
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object HaplogroupClassifier {

    suspend fun analyze(
        kitName: String,
        parsedData: ParsedDnaData
    ): DnaAnalysisResult = withContext(Dispatchers.Default) {
        // 1. Evaluate all known markers
        val evaluatedMarkers = evaluateMarkers(parsedData)

        // 2. Classify Paternal Lineage (Y-DNA)
        val yMarkers = evaluatedMarkers.filter { it.snp.lineageType == LineageType.PATERNAL_YDNA }
        val hasYData = parsedData.yDnaSnps > 0 || yMarkers.any { it.status != MarkerStatus.NO_CALL }
        val paternalLineage = if (hasYData) {
            classifyLineage(LineageType.PATERNAL_YDNA, yMarkers)
        } else {
            null
        }

        // 3. Classify Maternal Lineage (mtDNA)
        val mtMarkers = evaluatedMarkers.filter { it.snp.lineageType == LineageType.MATERNAL_MTDNA }
        val maternalLineage = classifyLineage(LineageType.MATERNAL_MTDNA, mtMarkers)

        DnaAnalysisResult(
            kitName = kitName,
            rawFileFormat = parsedData.format,
            totalSnpsParsed = parsedData.totalSnps,
            yDnaSnpsCount = parsedData.yDnaSnps,
            mtDnaSnpsCount = parsedData.mtDnaSnps,
            paternalLineage = paternalLineage,
            maternalLineage = maternalLineage,
            isMaleSample = paternalLineage != null && yMarkers.any { it.status == MarkerStatus.POSITIVE_DERIVED }
        )
    }

    private fun evaluateMarkers(parsedData: ParsedDnaData): List<EvaluatedMarker> {
        val result = mutableListOf<EvaluatedMarker>()

        for (snp in SnpReferenceDatabase.ALL_DEFINING_SNPS) {
            val rsidKey = snp.rsid.lowercase()
            val posKey = "${snp.chromosome.lowercase()}:${snp.position}"

            // Try to find genotype by rsid or by chr:pos
            val userGenotype = parsedData.snpByRsid[rsidKey]
                ?: parsedData.snpByPosition[posKey]
                ?: "--"

            val status = when {
                userGenotype == "--" || userGenotype.isEmpty() -> MarkerStatus.NO_CALL
                isGenotypeMatching(userGenotype, snp.derivedAllele) -> MarkerStatus.POSITIVE_DERIVED
                isGenotypeMatching(userGenotype, snp.ancestralAllele) -> MarkerStatus.NEGATIVE_ANCESTRAL
                else -> MarkerStatus.MISMATCH
            }

            val details = when (status) {
                MarkerStatus.POSITIVE_DERIVED -> "Derived mutation detected (${snp.derivedAllele}). Positive for ${snp.haplogroup}."
                MarkerStatus.NEGATIVE_ANCESTRAL -> "Ancestral allele observed (${snp.ancestralAllele}). Unmutated."
                MarkerStatus.NO_CALL -> "Marker uncalled or not covered in raw data."
                MarkerStatus.MISMATCH -> "Genotype '$userGenotype' differs from expected ancestral (${snp.ancestralAllele}) & derived (${snp.derivedAllele})."
            }

            result.add(EvaluatedMarker(snp, userGenotype, status, details))
        }

        return result
    }

    private fun isGenotypeMatching(userGenotype: String, targetAllele: String): Boolean {
        val u = userGenotype.uppercase()
        val t = targetAllele.uppercase()

        if (t == "INS" || t == "I") return u.contains("I") || u.contains("INS")
        if (t == "DEL" || t == "D") return u.contains("D") || u.contains("DEL")

        // For SNPs, user genotype can be single "A" or double "AA" or heterozygous "AG"
        return u.contains(t)
    }

    private fun classifyLineage(
        type: LineageType,
        markers: List<EvaluatedMarker>
    ): LineageAnalysis {
        val haplogroups = if (type == LineageType.PATERNAL_YDNA) {
            HaplogroupTreeData.Y_DNA_HAPLOGROUPS
        } else {
            HaplogroupTreeData.MT_DNA_HAPLOGROUPS
        }

        // Calculate scores for each candidate haplogroup
        val scoredHaplos = haplogroups.map { haplo ->
            val haploMarkers = markers.filter { m ->
                haplo.definingSnps.any { snpName ->
                    m.snp.name.equals(snpName, ignoreCase = true) ||
                    m.snp.rsid.equals(snpName, ignoreCase = true) ||
                    m.snp.haplogroup.equals(haplo.code, ignoreCase = true)
                }
            }

            val positives = haploMarkers.count { it.status == MarkerStatus.POSITIVE_DERIVED }
            val negatives = haploMarkers.count { it.status == MarkerStatus.NEGATIVE_ANCESTRAL }
            val total = haploMarkers.size

            HaploScore(
                haplogroup = haplo,
                positives = positives,
                negatives = negatives,
                totalMarkers = total,
                depth = calculateCladeDepth(haplo, haplogroups)
            )
        }

        // Find the best terminal haplogroup (highest positives and deepest clade)
        val bestCandidate = scoredHaplos
            .filter { it.positives > 0 }
            .sortedWith(
                compareByDescending<HaploScore> { it.positives }
                    .thenByDescending { it.depth }
                    .thenBy { it.negatives }
            )
            .firstOrNull()
            ?: scoredHaplos.firstOrNull()
            ?: HaploScore(haplogroups.first(), 0, 0, 0, 0)

        // Build tree path from Root to Terminal
        val treePath = buildLineagePath(bestCandidate.haplogroup, haplogroups)

        // Calculate confidence (higher if more defining SNPs positive, reduced by negative branch conflicts)
        val totalPos = markers.count { it.status == MarkerStatus.POSITIVE_DERIVED }
        val totalNeg = markers.count { it.status == MarkerStatus.NEGATIVE_ANCESTRAL }
        val confidence = if (bestCandidate.positives > 0) {
            val baseConfidence = when {
                bestCandidate.positives >= 3 -> 99
                bestCandidate.positives == 2 -> 96
                bestCandidate.positives == 1 -> 90
                else -> 75
            }
            (baseConfidence - (bestCandidate.negatives * 10)).coerceIn(50, 99)
        } else {
            30
        }

        return LineageAnalysis(
            lineageType = type,
            terminalHaplogroup = bestCandidate.haplogroup,
            confidenceScore = confidence,
            positiveCount = totalPos,
            negativeCount = totalNeg,
            totalTestedMarkers = markers.size,
            lineageTreePath = treePath,
            evaluatedMarkers = markers
        )
    }

    private fun calculateCladeDepth(
        haplo: HaplogroupDefinition,
        allHaplos: List<HaplogroupDefinition>
    ): Int {
        var depth = 1
        var current: HaplogroupDefinition? = haplo
        while (current?.parentClade != null) {
            depth++
            current = allHaplos.firstOrNull { it.code.equals(current?.parentClade, ignoreCase = true) }
        }
        return depth
    }

    private fun buildLineagePath(
        terminal: HaplogroupDefinition,
        allHaplos: List<HaplogroupDefinition>
    ): List<HaplogroupDefinition> {
        val path = mutableListOf<HaplogroupDefinition>()
        var current: HaplogroupDefinition? = terminal
        while (current != null) {
            path.add(0, current)
            val parentCode = current.parentClade
            current = if (parentCode != null) {
                allHaplos.firstOrNull { it.code.equals(parentCode, ignoreCase = true) }
            } else {
                null
            }
        }
        return path
    }

    private data class HaploScore(
        val haplogroup: HaplogroupDefinition,
        val positives: Int,
        val negatives: Int,
        val totalMarkers: Int,
        val depth: Int
    )
}

File: src/services/haplogroupClassifier.ts
import { ALL_DEFINING_SNPS } from '../data/snpDatabase';
import { Y_DNA_HAPLOGROUPS, MT_DNA_HAPLOGROUPS } from '../data/haplogroupTree';
import { ParsedDnaData } from './dnaParser';
import {
  DnaAnalysisResult,
  EvaluatedMarker,
  HaplogroupDefinition,
  LineageAnalysis,
  LineageType,
  MarkerStatus
} from '../types/haplogroup';

interface HaploScore {
  haplogroup: HaplogroupDefinition;
  positives: number;
  negatives: number;
  totalMarkers: number;
  depth: number;
}

export class HaplogroupClassifier {
  public static analyze(kitName: string, parsedData: ParsedDnaData): DnaAnalysisResult {
    // 1. Evaluate all defining SNPs against the parsed genomic data
    const evaluatedMarkers = this.evaluateMarkers(parsedData);

    // 2. Classify Paternal Lineage (Y-DNA)
    const yMarkers = evaluatedMarkers.filter(m => m.snp.lineageType === 'PATERNAL_YDNA');
    const hasYData = parsedData.yDnaSnps > 0 || yMarkers.some(m => m.status !== 'NO_CALL');
    const paternalLineage = hasYData ? this.classifyLineage('PATERNAL_YDNA', yMarkers) : null;

    // 3. Classify Maternal Lineage (mtDNA)
    const mtMarkers = evaluatedMarkers.filter(m => m.snp.lineageType === 'MATERNAL_MTDNA');
    const maternalLineage = this.classifyLineage('MATERNAL_MTDNA', mtMarkers);

    const isMale = paternalLineage !== null && yMarkers.some(m => m.status === 'POSITIVE_DERIVED');

    return {
      id: `kit_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`,
      kitName,
      timestamp: Date.now(),
      rawFileFormat: parsedData.format,
      totalSnpsParsed: parsedData.totalSnps,
      yDnaSnpsCount: parsedData.yDnaSnps,
      mtDnaSnpsCount: parsedData.mtDnaSnps,
      paternalLineage,
      maternalLineage,
      isMaleSample: isMale
    };
  }

  private static evaluateMarkers(parsedData: ParsedDnaData): EvaluatedMarker[] {
    const result: EvaluatedMarker[] = [];

    for (const snp of ALL_DEFINING_SNPS) {
      const rsidKey = snp.rsid.toLowerCase();
      const posKey = `${snp.chromosome.toLowerCase()}:${snp.position}`;

      const userGenotype = parsedData.snpByRsid[rsidKey] || parsedData.snpByPosition[posKey] || '--';

      let status: MarkerStatus = 'NO_CALL';
      let details = '';

      if (userGenotype === '--' || !userGenotype) {
        status = 'NO_CALL';
        details = 'Marker uncalled or not covered in raw data.';
      } else if (this.isGenotypeMatching(userGenotype, snp.derivedAllele)) {
        status = 'POSITIVE_DERIVED';
        details = `Derived mutation detected (${snp.derivedAllele}). Positive for clade ${snp.haplogroup}.`;
      } else if (this.isGenotypeMatching(userGenotype, snp.ancestralAllele)) {
        status = 'NEGATIVE_ANCESTRAL';
        details = `Ancestral allele observed (${snp.ancestralAllele}). Unmutated.`;
      } else {
        status = 'MISMATCH';
        details = `Genotype '${userGenotype}' differs from expected ancestral (${snp.ancestralAllele}) & derived (${snp.derivedAllele}).`;
      }

      result.push({
        snp,
        userGenotype,
        status,
        details
      });
    }

    return result;
  }

  private static isGenotypeMatching(userGenotype: string, targetAllele: string): boolean {
    const u = userGenotype.toUpperCase();
    const t = targetAllele.toUpperCase();

    if (t === 'INS' || t === 'I') return u.includes('I') || u.includes('INS');
    if (t === 'DEL' || t === 'D') return u.includes('D') || u.includes('DEL');

    return u.includes(t);
  }

  private static classifyLineage(type: LineageType, markers: EvaluatedMarker[]): LineageAnalysis {
    const haplogroups = type === 'PATERNAL_YDNA' ? Y_DNA_HAPLOGROUPS : MT_DNA_HAPLOGROUPS;

    const scoredHaplos: HaploScore[] = haplogroups.map(haplo => {
      const haploMarkers = markers.filter(m => 
        haplo.definingSnps.some(s => 
          s.toLowerCase() === m.snp.name.toLowerCase() ||
          s.toLowerCase() === m.snp.rsid.toLowerCase() ||
          s.toLowerCase() === m.snp.haplogroup.toLowerCase() ||
          haplo.code.toLowerCase() === m.snp.haplogroup.toLowerCase()
        )
      );

      const positives = haploMarkers.filter(m => m.status === 'POSITIVE_DERIVED').length;
      const negatives = haploMarkers.filter(m => m.status === 'NEGATIVE_ANCESTRAL').length;

      return {
        haplogroup: haplo,
        positives,
        negatives,
        totalMarkers: haploMarkers.length,
        depth: this.calculateCladeDepth(haplo, haplogroups)
      };
    });

    // Best candidate: max positives -> deepest depth -> min negatives
    const candidates = scoredHaplos.filter(h => h.positives > 0);
    candidates.sort((a, b) => {
      if (b.positives !== a.positives) return b.positives - a.positives;
      if (b.depth !== a.depth) return b.depth - a.depth;
      return a.negatives - b.negatives;
    });

    const bestCandidate = candidates[0] || scoredHaplos[0] || {
      haplogroup: haplogroups[0],
      positives: 0,
      negatives: 0,
      totalMarkers: 0,
      depth: 1
    };

    const treePath = this.buildLineagePath(bestCandidate.haplogroup, haplogroups);

    const totalPos = markers.filter(m => m.status === 'POSITIVE_DERIVED').length;
    const totalNeg = markers.filter(m => m.status === 'NEGATIVE_ANCESTRAL').length;

    let confidence = 30;
    if (bestCandidate.positives > 0) {
      const baseConfidence = bestCandidate.positives >= 3 ? 99 : (bestCandidate.positives === 2 ? 96 : 90);
      confidence = Math.max(50, Math.min(99, baseConfidence - (bestCandidate.negatives * 10)));
    }

    return {
      lineageType: type,
      terminalHaplogroup: bestCandidate.haplogroup,
      confidenceScore: confidence,
      positiveCount: totalPos,
      negativeCount: totalNeg,
      totalTestedMarkers: markers.length,
      lineageTreePath: treePath,
      evaluatedMarkers: markers
    };
  }

  private static calculateCladeDepth(haplo: HaplogroupDefinition, allHaplos: HaplogroupDefinition[]): number {
    let depth = 1;
    let current: HaplogroupDefinition | undefined = haplo;
    while (current?.parentClade) {
      depth++;
      const parentCode: string = current.parentClade;
      current = allHaplos.find(h => h.code.toLowerCase() === parentCode.toLowerCase());
    }
    return depth;
  }

  private static buildLineagePath(terminal: HaplogroupDefinition, allHaplos: HaplogroupDefinition[]): HaplogroupDefinition[] {
    const path: HaplogroupDefinition[] = [];
    let current: HaplogroupDefinition | undefined = terminal;
    while (current) {
      path.unshift(current);
      if (!current.parentClade) break;
      const parentCode: string = current.parentClade;
      current = allHaplos.find(h => h.code.toLowerCase() === parentCode.toLowerCase());
    }
    return path;
  }
}
